Add Python 3.4/5/6 and PyPy support.

jamadden opened the pull request:
Coverage isn't perfect but its pretty good.

This blocks zope.app.rotterdam, which blocks zope.app.component.

Test coverage isn't 100% but it's pretty good at 97%.

There was actually almost nothing to do here aside from implements to implementer and a few doctest cleanups.

I did [discover a bug](https://github.com/zopefoundation/zope.configuration/issues/17) in `zope.configuration.fields.MessageID` but the workaround was easy and not terribly ugly.

jamadden commented:
I was actually able to get zope.app.rotterdam to work independent of this; it's not needed for anything other than interfaces and doesn't have to be configured for the tests to work. See zopefoundation/zope.app.rotterdam#1.

icemac commented on src/zope/app/form/testing.py:
```diff
@@ -19,9 +19,44 @@
+
+stypes = list, tuple
+def provideAdapter(required, provided, factory, name='', with_=(), **kw):
+    if with_ is None and kw.has_key('with'): # pragma: no cover
```
Replace `kw.has_key('with')` by `'with' in kw` to be Python 3 compatible?
Or remove it because it seems to be there for Python 2.4 compatibility (before `with` statements got introduced).

jamadden replied:
It's a no-cover line so we can probably just remove the whole thing. Not sure why I left it if it's not being used.

icemac commented on src/zope/app/form/tests/test_utility.py:
```diff
@@ -1029,114 +1036,127 @@ def test_typical(self):
-        
-            >>> keys = result.keys(); keys.sort(); keys
+
+            >>> keys = list(result.keys()); keys.sort(); keys
```
How about `sorted(result.keys())`? (Tried it in Python 2 and 3, seems to work.)

jamadden replied:
Sure, that's simpler. I just did the minimal change to make it work :)

icemac commented on src/zope/app/form/tests/test_utility.py:
```diff
@@ -1029,114 +1036,127 @@ def test_typical(self):
+    (re.compile("u('.*?')"), r"\1"),
+    (re.compile('u(".*?")'), r"\1"),
+    # Python 3 adds module name to exceptions.
+    (re.compile('zope.interface.exceptions.Invalid'), 'Invalid'),
```
`IGNORE_EXCEPTION_MODULE_IN_PYTHON2` here as stated above.

jamadden commented on MANIFEST.in:
```diff
@@ -1,9 +1,12 @@
 include *.txt
+include *.rst
+include .coveragerc
+include tox.ini
```
Downstream distributors (e.g., linux distros, probably conda, etc) do run tests from sdists. I've had a number of complaints from gevent distributors about the test suite it ships.

mgedmin replied:
It depends on whether you think sdists are meant for machines to build, or for people to use.  While most people will use git clones, sometimes you need to use an sdist (e.g. you need to backport a quick fix onto a pinned version, and it's simpler to grab the sdist from PyPI than to hunt down the exact branch/tag/commit from whatever repo upstream uses).

In those circumstances it's helpful to provide files like .coveragerc or tox.ini (or even .travis.yml) so people can look at them and see what's the procedure to run tests for this particular piece of software.

jamadden replied:
@mgedmin makes a good point. I often find myself looking at tox.ini/.travis.yml to figure out test procedures. I wish I'd been including .travis.yml in the MANIFEST.in all this time (I didn't because I check-manifest didn't complain...I guess it ignores dotfiles, although that's not how its documented).

mgedmin replied:
check-manifest doesn't complain about `.travis.yml` because [people asked for that](https://github.com/mgedmin/check-manifest/issues/10#issuecomment-18551337).  There's no generic rule for ignoring dotfiles, just a [default ignore list](https://github.com/mgedmin/check-manifest/blob/58110391c941c7f6eeabc16e5dadd9def6d1e707/check_manifest.py#L509).

jamadden replied:
Ah, the version of the ignore list [documented on PyPI](https://pypi.python.org/pypi/check-manifest) doesn't include '.travis.yml'.

jamadden commented on src/zope/app/form/browser/i18n.rst:
```diff
@@ -8,15 +8,15 @@ i18n messages in the schema.
 Let's take this simple add form...
 
-  >>> print http(r"""
+#>>> print(http(r"""
```
Whoops. Thanks for catching that.

jamadden commented on src/zope/app/form/browser/tests/test_form.py:
```diff
@@ -21,19 +21,31 @@
+    (re.compile("u('.*?')"), r"\1"),
+    (re.compile('u(".*?")'), r"\1"),
+    # Python 3 adds module name to exceptions.
+    (re.compile('zope.configuration.xmlconfig.ZopeXMLConfigurationError'), 'ZopeXMLConfigurationError'),
```
Interesting, I wasn't aware of that flag. 

Using it will require more changes to the doctests, though, to use fully-qualified dotted names in all cases. That might or might not be a good thing. Unless you object, I would just as soon leave things the way they are for right now.

jamadden commented on .coveragerc:
```diff
@@ -0,0 +1,8 @@
+exclude_lines =
+    pragma: no cover
+    raise NotImplementedError
+    if __name__ == '__main__':
```
I guess we could do that. I'm personally pretty used to .coveragerc though and would prefer not to; one minor reason is that one or the other of the files may be customized in a series of projects while the other is identical, so it's easier to copy around if the files are separate (I admit that's not an especially great reason).

jamadden commented on src/zope/app/form/browser/tests/test_form.py:
```diff
@@ -21,19 +21,31 @@
+import re
+checker = renormalizing.RENormalizing([
+    (re.compile("u('.*?')"), r"\1"),
+    (re.compile('u(".*?")'), r"\1"),
```
Yes 😦 The regex didn't match inside the list/tuple, it only handles an entire line. It was easier to just use print those than to fuss with the regex.

mgedmin replied:
Ah, the regex is too greedy!  Replacing `.*` with `[^']*` would've probably worked (does Python have regex syntax for a non-greedy `*` a la Perl's `*?`?) , but the print solution is fine too.

jamadden commented on src/zope/app/form/browser/metaconfigure.py:
```diff
@@ -62,6 +63,20 @@ def __init__(self, _context, **kwargs):
+                    and isinstance(value.domain, bytes)):
+                    domain = value.domain.decode('ascii')
+                    factory = zope.i18nmessageid.MessageFactory(domain)
+                    value = factory(str(value), value.default)
```
Thanks for reminding me, I didn't have one filed yet but I do now.

jamadden commented on src/zope/app/form/browser/tests/test_functional_editview.py:
```diff
@@ -19,31 +19,53 @@
+
+    def publish(self, path, basic=None, form=None, headers=None):
+        if basic:
+            self._testapp.authorization = ('Basic', tuple(basic.split(':')))
```
Right! The username and password are hardcoded all over the place.

jamadden commented on src/zope/app/form/browser/tests/test_functional_i18n.py:
```diff
@@ -62,23 +64,60 @@ class IFieldContent(Interface):
+class _FakeResponse(FakeResponse):
+    if str is bytes:
+        # py2 has a bug, assuming headers are in unicode already, or
+        # are decodable implicitly to ascii
```
I don't have one filed there yet. But I can at least leave a version number for now.

jamadden commented on src/zope/app/form/browser/widgets.rst:
```diff
@@ -13,56 +13,60 @@ description
+    ... class Poll(Persistent, object):
+    ...
+    ...
+    ...  def getResponse(self, option):
```
The blank line is a result of moving around the implements/implementer. The small indentation is a result of doing the bare minimum to turn this code block into a working doctest. I don't know of an editor that correctly and context-sensitively automatically handles indents inside a doctest inside a rst file, so I wind up counting spaces which is very tedious.

mgedmin replied:
vim FTW :)

jamadden replied:
No way, vim does that? Out of the box or with custom plugins/configuration? 

I always figured it was a hard problem. But if vim can do it, emacs must be able to as well...I have some research to do.

mgedmin replied:
There's some out-of-the-box support (set softtabstop=4 and hitting <tab>/<backspace> indents/dedents the rest of the line to the right mutliple of four) and some plugins (to get Python syntax higlighting in doctests in ReStructuredText files).  I also like visual block mode for adding the doctest prefix to a bunch of lines in one go, when you move existing code into a doctest.

IIRC I once tried very hard to make autoindentation work correctly (i.e. insert a newline, four spaces, three dots, a space, and then the right number of spaces for Python code), but gave up without achieving success.

jamadden commented on src/zope/app/form/browser/widgets.rst:
```diff
@@ -13,56 +13,60 @@ description
+    ...        for option in self._options:
+    ...            self._responses[option.label] = 0
+    ...
+    ...  options = property(get_options, set_options, None, 'fiddle options')
```
We could, but that would mean more futzing with lines in a docstring in a rst...

mgedmin replied:
FWIW my reviews are generally "here's how I think you can make the code better", and not "you must do this before I'll allow a merge".  I'm bad at stating that explicitly (and I'm bad at using [positive language](https://www.pagerduty.com/blog/better-code-reviews-empathy/) in my reviews, despite my frequent resolutions to get better at it).

jamadden commented on src/zope/app/form/tests/test_bwc_imports.py:
```diff
@@ -0,0 +1,39 @@
+                               ('browser.exception', 'WidgetInputErrorView'),
+                               ('browser.boolwidgets', 'CheckBoxWidget'),
+                               ('browser.itemswidgets', 'ChoiceDisplayWidget')):
+        locals()['test_' + mod_name] = _make_import_test(mod_name, attrname)
```
Using `locals()` in this way in a class body is fairly common. gevent does it, as does Acquisition. Standard library modules like pyexpat do it, and the stdlib test suite does exactly this quite a bit (test_scope, test_os, test_compilerall, etc) as does setuptools. The rpython implementation of PyPy (which runs under CPython) uses it extensively.

I will still change it if you feel strongly.

mgedmin replied:
I've never seen it before!   I'm fine with it, given that popular libraries rely on it.

(I still prefer my FunctionTestCase solution, but I'll not insist.)

jamadden commented:
The MessageID bug has been fixed. Do we want to wait for a new zope.configuration and code a dependency on that version, or still have a workaround but leave the zope.configuration version unpinned?

jamadden commented:
I think I've addressed all the comments, and the two bugs we had workarounds for have been fixed, released, and workarounds removed. Since there are two approvals, I'll go ahead and merge. Thanks everyone!

Could we please get a PyPI release when convenient?

mgedmin commented:
You now have PyPI rights.

jamadden commented:
Thanks! Released to PyPI.
